### PR TITLE
common : refactor common_sampler + grammar logic changes

### PR DESCRIPTION
**Pull request by ggerganov**
from #17004

Extracting some refactoring portions from #17004 to make the review easier:

- Simplify and make safer the management of llama objects (samplers, contexts, model)
- The `common_init_result` now also owns the sampler chains constructed during `common_init_from_params()`
- The sampler chains of `common_init_result` are constructed before the model and the context - we will need this for #17004 in order to optionally pass the samplers during the construction of the context

---

ref https://github.com/ggml-org/llama.cpp/pull/17750#issuecomment-3619743766

Another change related to the grammar logic (the explanation is in the referenced comment):

- No longer maintain a separate sampler chain for the grammar
- Merge the grammar into the main `common_sampler` chain
- **The grammar is now always applied first to the raw logits, before the rest of the samplers**

The main reason for this change is to make the integration of #17004 compatible with grammar usage and to simplify the logic for handling the grammar when it is present. The main concern is that this will likely hurt the performance when grammar sampling is involved, since we no longer do the ["rejection sampling" trick](https://github.com/ggml-org/llama.cpp/blob/c6f6e4f96a7f7bce49f5c21d19ee69fb8b72f84d/common/sampling.cpp#L406-L428). I think it's better to put effort to optimize the performance of the grammar in general so we don't need to do the trick at all.

### CODE REVIEW

File: examples/batched/batched.cpp [hunk 1]
```diff
@@ -173,7 +178,7 @@ int main(int argc, char ** argv) {
                 continue;
             }
 
-            const llama_token new_token_id = llama_sampler_sample(smpl, ctx, i_batch[i]);
+            const llama_token new_token_id = llama_sampler_sample(sampler_configs[i], ctx, i_batch[i]);
```
Comment (danbev) [inline examples/batched/batched.cpp]: I looks like `sampler_configs` is not populated and will currently cause a segment fault:
```console
Thread 1 "llama-batched" received signal SIGSEGV, Segmentation fault.
0x000055555566ad40 in main (argc=12, argv=0x7fffffffd688) at /home/danbev/work/ai/llama.cpp/examples/batched/batched.cpp:181
181	            const llama_token new_token_id = llama_sampler_sample(sampler_configs[i], ctx, i_batch[i]);
```

File: examples/batched/batched.cpp [hunk 2]
```diff
@@ -64,17 +65,21 @@ int main(int argc, char ** argv) {
     ctx_params.n_ctx   = n_kv_req;
     ctx_params.n_batch = std::max(n_predict, n_parallel);
 
-    llama_context * ctx = llama_init_from_model(model, ctx_params);
-
     auto sparams = llama_sampler_chain_default_params();
     sparams.no_perf = false;
 
-    llama_sampler * smpl = llama_sampler_chain_init(sparams);
+    std::vector<llama_sampler *> sampler_configs;
```
Comment (danbev) [inline examples/batched/batched.cpp]: Perhaps this should be named `samplers` instead?

File: examples/batched/batched.cpp [hunk 3]
```diff
@@ -64,17 +65,21 @@ int main(int argc, char ** argv) {
...
-    llama_sampler_chain_add(smpl, llama_sampler_init_top_p(params.sampling.top_p, params.sampling.min_keep));
-    llama_sampler_chain_add(smpl, llama_sampler_init_temp (params.sampling.temp));
-    llama_sampler_chain_add(smpl, llama_sampler_init_dist (params.sampling.seed));
+    for (int32_t i = 0; i < n_parallel; ++i) {
+        llama_sampler * smpl = llama_sampler_chain_init(sparams);
+
+        llama_sampler_chain_add(smpl, llama_sampler_init_top_k(params.sampling.top_k));
+        llama_sampler_chain_add(smpl, llama_sampler_init_top_p(params.sampling.top_p, params.sampling.min_keep));
+        llama_sampler_chain_add(smpl, llama_sampler_init_temp (params.sampling.temp));
+        llama_sampler_chain_add(smpl, llama_sampler_init_dist (params.sampling.seed));
```
Comment (danbev) [inline examples/batched/batched.cpp:79]: Should the samplers also be added to the samplers:
```suggestion
        llama_sampler_chain_add(smpl, llama_sampler_init_dist (params.sampling.seed));
        sampler_configs.push_back(smpl);
```

Comment (ggerganov) [reply examples/batched/batched.cpp:79]: Should be fixed now

**Review by danbev — APPROVED** [review]
